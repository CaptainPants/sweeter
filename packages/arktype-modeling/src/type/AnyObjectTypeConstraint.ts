import { ObjectType } from 'arktype/internal/methods/object.ts';

export type AnyObjectTypeConstraint = ObjectType<object, any>;
