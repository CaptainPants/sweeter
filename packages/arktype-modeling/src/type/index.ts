export * from './types.js';
export * from './introspect/index.js';
