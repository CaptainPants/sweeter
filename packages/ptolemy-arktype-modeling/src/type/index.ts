export * from './introspect/index.js';
export * from './types.js';
