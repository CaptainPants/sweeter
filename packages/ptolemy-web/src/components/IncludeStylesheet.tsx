/* @jsxImportSource .. */

import {
    $derived,
    $val,
    type Component,
    type MightBeSignal,
    type Signal,
} from '@serpentis/ptolemy-core';
import { arrayExcept } from '@serpentis/ptolemy-utilities';

import { getWebRuntime } from '../runtime/getWebRuntime.js';
import {
    type AbstractGlobalCssStylesheet,
    type StylesheetInclude,
} from '../styles/index.js';

export interface IncludeStylesheetProps {
    stylesheet: StylesheetInclude;
};

export const IncludeStylesheet: Component<IncludeStylesheetProps> = (
    { stylesheet },
    init,
) => {
    const webRuntime = getWebRuntime();

    let previousTime: AbstractGlobalCssStylesheet[] | undefined;

    const flattened = flatten(stylesheet);

    // Remember that this is called on mount as well as when the subscribed signals change.
    init.trackSignals([flattened], ([thisTime]) => {
        if (previousTime) {
            const added = arrayExcept(thisTime, previousTime);
            const removed = arrayExcept(previousTime, thisTime);

            for (const addedItem of added) {
                webRuntime.addStylesheet(addedItem);
            }
            for (const removedItem of removed) {
                webRuntime.removeStylesheet(removedItem);
            }
        } else {
            for (const addedItem of thisTime) {
                webRuntime.addStylesheet(addedItem);
            }
        }

        previousTime = thisTime;
    });

    init.onUnMount(() => {
        if (previousTime) {
            for (const removedItem of previousTime) {
                webRuntime.removeStylesheet(removedItem);
            }
        }
    });

    return <></>;
};

function flatten(
    value: MightBeSignal<StylesheetInclude>,
): Signal<AbstractGlobalCssStylesheet[]> {
    return $derived(() => {
        value = $val(value);

        if (Array.isArray(value)) {
            return value;
        } else {
            return [value];
        }
    });
}
