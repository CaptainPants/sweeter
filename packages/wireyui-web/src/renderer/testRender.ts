import { WebRenderer } from './WebRenderer.js';

export interface TestRenderResult {
    readonly nodes: NodeListOf<ChildNode>;
    readonly getHTML: () => string;
    readonly dispose: () => void;
}

/**
 * Convenience function for rendering isolated JSX chunks for unit tests
 * @param callback
 * @returns
 */
export function testRender(callback: () => JSX.Element): TestRenderResult {
    const renderer = new WebRenderer();

    const root = document.createElement('div');

    const dispose = renderer.start(root, callback);

    return {
        nodes: root.childNodes,
        getHTML: () => root.innerHTML,
        dispose,
    };
}
