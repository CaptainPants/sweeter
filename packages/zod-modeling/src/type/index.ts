export * from './is/index.js';
export * from './unwrap.js';