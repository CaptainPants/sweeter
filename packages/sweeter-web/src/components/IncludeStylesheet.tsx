/* @jsxImportSource .. */

import type { Component } from '@captainpants/sweeter-core';
import type { GlobalCssStylesheet } from '../styles/index.js';
import { getWebRuntime } from '../runtime/getWebRuntime.js';

export interface IncludeStylesheetProps {
    stylesheet: GlobalCssStylesheet;
}

export const IncludeStylesheet: Component<IncludeStylesheetProps> = (
    { stylesheet },
    init,
) => {
    const runtime = getWebRuntime();

    init.subscribeToChanges(
        [stylesheet],
        ([stylesheet]) => {
            if (stylesheet) {
                const remove = runtime.addStylesheet(stylesheet);

                return remove;
            }

            return undefined;
        },
        true,
    );

    return <></>;
};
