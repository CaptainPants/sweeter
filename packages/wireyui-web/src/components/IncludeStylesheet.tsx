/* @jsxImportSource .. */

import type { Component } from '@captainpants/wireyui-core';
import type { AbstractGlobalCssStylesheet } from '../styles/index.js';
import type { DocumentStylesheetHandle } from '../runtime/WebRuntimeContext.js';
import { WebRuntimeContext } from '../runtime/WebRuntimeContext.js';

export interface IncludeStylesheetProps {
    stylesheet: AbstractGlobalCssStylesheet;
}

export const IncludeStylesheet: Component<IncludeStylesheetProps> = (
    { stylesheet },
    init,
) => {
    const runtime = init.getContext(WebRuntimeContext);

    let handle: DocumentStylesheetHandle | undefined;

    init.subscribeToChanges(
        [stylesheet],
        ([stylesheet]) => {
            if (handle) {
                if (stylesheet) {
                    handle.update(stylesheet);
                } else {
                    handle.remove();
                }
            } else {
                if (stylesheet) {
                    handle = runtime.addStylesheet(stylesheet);
                }
            }
            return;
        },
        true,
    );

    init.onUnMount(() => {
        handle?.remove();
    });

    return <></>;
};
