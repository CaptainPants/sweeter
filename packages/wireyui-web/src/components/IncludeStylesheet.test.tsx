/* @jsxImportSource .. */

import type { Component } from '@captainpants/wireyui-core';
import { GlobalCss, GlobalCssClass, StyleHooks } from '../styles/index.js';
import { testRender } from '../test/testRender.js';
import { IncludeStylesheet } from './IncludeStylesheet.js';

const example = new GlobalCss({
    id: 'Something',
    content: `
        body { background-color: red; }
    `,
});

it('Global stylesheet added', () => {
    const res = testRender(() => {
        return <IncludeStylesheet stylesheet={example} />;
    });

    expect(document.head.innerHTML).toMatchSnapshot();

    res.dispose();
});

const componentStylesheet = new GlobalCssClass({
    id: 'Component1',
    content: `
        color: blue;
    `,
});

const Component1: Component = (props, init) => {
    const styleHooks = init(StyleHooks);

    return (
        <>
            <IncludeStylesheet stylesheet={componentStylesheet} />
            <div class={styleHooks.getClassName(componentStylesheet)}>
                This is an example
            </div>
        </>
    );
};

it('Global style for component', () => {
    const res = testRender(() => {
        return <Component1 />;
    });

    expect(document.head.innerHTML).toMatchSnapshot();

    expect(res.getHTML()).toMatchSnapshot();

    res.dispose();
});
