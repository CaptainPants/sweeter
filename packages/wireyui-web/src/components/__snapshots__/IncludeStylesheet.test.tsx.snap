// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global style for component 1`] = `
"<style data-id="Component1">._glbl0_Component1
{

        color: blue;
    
}</style>"
`;

exports[`Global style for component 2`] = `"<!--Hooks(IncludeStylesheet): Mount, UnMount--><div class="_glbl0_Component1">This is an example</div>"`;

exports[`Global stylesheet added 1`] = `
"<style data-id="Something">
        body { background-color: red; }
    </style>"
`;
